bindings/rust/src/lib.rs

```diff
@@ -40,14 +40,66 @@ mod sys;
 use std::ffi::CString;
 use std::ptr::NonNull;
 
+/// A safe container for handling the low-level FizzyError struct.
+struct CErrorContainer(Box<sys::FizzyError>);
+
+impl CErrorContainer {
+    /// Create a safe, boxed, and zero initialised container.
+    fn new() -> Self {
+        CErrorContainer {
+            0: Box::new(sys::FizzyError {
+                code: 0,
+                message: [0i8; 256],
+            }),
+        }
+    }
+
+    /// Return a pointer passable to low-level functions (validate, parse, instantiate).
+    ///
+    /// # Safety
+    /// The returned pointer is still onwed by this struct, and thus not valid after this struct goes out of scope.
+    unsafe fn as_mut_ptr(&mut self) -> *mut sys::FizzyError {
+        &mut *self.0
+    }
+
+    /// Return the underlying error code.
+    // TODO: represent the errors as proper Rust enums
+    fn code(&self) -> u32 {
+        self.0.code
+    }
+
+    /// Return an owned String copy of the underlying message.
+    fn message(&self) -> String {
+        unsafe {
+            std::ffi::CStr::from_ptr(self.0.message.as_ptr())
+                .to_string_lossy()
+                .into_owned()
+        }
+    }
+}
+
+impl std::fmt::Display for CErrorContainer {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{} [{}]", self.code(), self.message())
+    }
+}
+
 /// Parse and validate the input according to WebAssembly 1.0 rules. Returns true if the supplied input is valid.
-pub fn validate<T: AsRef<[u8]>>(input: T) -> bool {
-    unsafe {
+pub fn validate<T: AsRef<[u8]>>(input: T) -> Result<(), String> {
+    let mut err = CErrorContainer::new();
+    let ret = unsafe {
         sys::fizzy_validate(
             input.as_ref().as_ptr(),
             input.as_ref().len(),
-            std::ptr::null_mut(),
+            err.as_mut_ptr(),
         )
+    };
+    if ret {
+        debug_assert!(err.code() == 0);
```

Comment:
These assertions show the current error reporting system is not prefect, as there are two ways to tell failure, error code or the return value.

Reply:
For validate theoretically we could return error code instead of `bool` (and optional error message), but for parse it seems inevitable: we have to return a pointer to module / instance in any case, and when we also return an error code this pointer is going to be NULL.

Reply:
Do we want to add assertions into the capi for this?

Reply:
> Do we want to add assertions into the capi for this? Or at least extend the tests? And document it?

Tests already have checks for both return value and error for both failure and success cases.

Assertions in C API - not sure where and which ones (it seems kind of obvious there that error is set when `false` or `nullptr` is returned)

Document it - something like`Returned error code is FIZZY_SUCCESS iff return value is true`? Sounds somewhat obvious, too...

Reply:
I guess documentation is fine just saying that `error code must be FIZZY_SUCCESS` if a pointer is returned and `must NOT be FIZZY_SUCCESS` otherwise?